### PR TITLE
Private message plugin new API implementation.

### PR DESCRIPTION
**Pull request by vpavliv**

### CODE REVIEW

File: libraries/plugins/private_message/include/golos/plugins/private_message/private_message_plugin.hpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+            DEFINE_API_ARGS(get_inbox,  json_rpc::msg_pack, inbox_r)
+            DEFINE_API_ARGS(get_outbox, json_rpc::msg_pack, outbox_r)
+
+            /**
+             *   This plugin scans the blockchain for custom operations containing a valid message and authorized
+             *   by the posting key.
+             *
+             */
+            class private_message_plugin : public appbase::plugin<private_message_plugin> {
```
Comment (kotbegemot) [inline libraries/plugins/private_message/include/golos/plugins/private_message/private_message_plugin.hpp]: where is the final modifier?

File: libraries/plugins/private_message/include/golos/plugins/private_message/private_message_plugin.hpp [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            public:
+
+                APPBASE_PLUGIN_REQUIRES((json_rpc::plugin))
+
+                private_message_plugin();
+
+                virtual ~private_message_plugin();
```
Comment (kotbegemot) [inline libraries/plugins/private_message/include/golos/plugins/private_message/private_message_plugin.hpp]: Why do you need a virtual ?

File: libraries/plugins/private_message/include/golos/plugins/private_message/private_message_plugin.hpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+                virtual void set_program_options(
```
Comment (kotbegemot) [inline libraries/plugins/private_message/include/golos/plugins/private_message/private_message_plugin.hpp]: Why do you need a virtual ?

File: libraries/plugins/private_message/include/golos/plugins/private_message/private_message_plugin.hpp [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                        boost::program_options::options_description &cli,
+                        boost::program_options::options_description &cfg) override;
+
+                virtual void plugin_initialize(const boost::program_options::variables_map &options) override;
```
Comment (kotbegemot) [inline libraries/plugins/private_message/include/golos/plugins/private_message/private_message_plugin.hpp]: Why do you need a virtual ?

File: libraries/plugins/private_message/include/golos/plugins/private_message/private_message_plugin.hpp [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+                virtual void plugin_startup() override;
```
Comment (kotbegemot) [inline libraries/plugins/private_message/include/golos/plugins/private_message/private_message_plugin.hpp]: Why do you need a virtual ?

File: libraries/plugins/private_message/include/golos/plugins/private_message/private_message_plugin.hpp [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+                virtual void plugin_shutdown() override;
```
Comment (kotbegemot) [inline libraries/plugins/private_message/include/golos/plugins/private_message/private_message_plugin.hpp]: Why do you need a virtual ?

File: libraries/plugins/private_message/private_message_plugin.cpp [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+//
+
+namespace golos {
+    namespace plugins {
+        namespace private_message {
+
+            class private_message_plugin::private_message_plugin_impl {
+            public:
+                private_message_plugin_impl(private_message_plugin &_plugin)
+                        : _self(_plugin) ,
```
Comment (kotbegemot) [inline libraries/plugins/private_message/private_message_plugin.cpp]: What for ?

File: libraries/plugins/private_message/private_message_plugin.cpp [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+                        pmo.receive_time = d.head_block_time();
+                        pmo.encrypted_message.resize(pm.encrypted_message.size());
+                        std::copy(pm.encrypted_message.begin(), pm.encrypted_message.end(),
+                                  pmo.encrypted_message.begin());
+                    });
+                }
+            }
+
+            private_message_plugin::private_message_plugin()
+                    : my(new private_message_plugin::private_message_plugin_impl(*this)) {
```
Comment (kotbegemot) [inline libraries/plugins/private_message/private_message_plugin.cpp]: Plugin  has its custom life cycle . Not use RAII.

Comment (vpavliv) [reply libraries/plugins/private_message/private_message_plugin.cpp]: In the code snippet above we use raii technique for impl. Impl does not exist out of plugin class. I do not quite understand the issue.

File: libraries/plugins/private_message/private_message_plugin.cpp (hunk 7 above)
Comment (vpavliv) [reply libraries/plugins/private_message/private_message_plugin.cpp]: What do you mean?

File: libraries/plugins/private_message/include/golos/plugins/private_message/private_message_plugin.hpp (hunk 1 above)
Comment (vpavliv) [reply libraries/plugins/private_message/include/golos/plugins/private_message/private_message_plugin.hpp]: In my next commit ;)

File: libraries/plugins/private_message/private_message_plugin.cpp (hunk 8 above)
Comment (vpavliv) [reply libraries/plugins/private_message/private_message_plugin.cpp]: Got it. Fixed.

**Pull request merged**